### ISSUE TITLE
DataPostprocessor fails with std::complex Vector

### ISSUE THREAD

**Issue by dangars** [dealii/dealii#8625]
Following up on #8617, in step-68 I calculate the eigenmodes of an electromagnetic cavity. The variable is the complex vector `electric_field`. To calculate `h_field= i*curl(E)/(omega*mu)` I use `DataPostprocessor` which seems to work fine except that it repeats names.

The problem is that `h_field` and `electric_field` are std::complex vectors. In the case of `electric_field` I can choose `DataComponentInterpretation::component_is_scalar` then give the appropriate names. In the case of `h_field` I don't have the option to give the names, therefore the names of `h_field` will be repeated. `h_field` has 6 fields 3x2, 3 for the dimensions a 2 for real and imaginary.

This is the output of the debugger for the `VTK` names, as you can see the names are repeated for `h_field`, therefore the program crashes.
```
$1 = std::vector of length 12, capacity 16 = {"electric_field_x_re",
"electric_field_x_im", "electric_field_y_re", "electric_field_y_im",
"electric_field_z_re", "electric_field_z_im", "h_field_re", "h_field_im",
"h_field_re", "h_field_im", "h_field_re", "h_field_im"}
```

I use the following code:

In the output function:
``` c++
HfieldPostprocessor<dim> h_field;
data_out.add_data_vector(locally_relevant_eigenfunctions, h_field);
```

The class code:
```c++
template <int dim> class HfieldPostprocessor : public DataPostprocessorVector<dim> {
public:
  HfieldPostprocessor() : DataPostprocessorVector<dim>("h_field", update_gradients) {}
  virtual void evaluate_vector_field(const DataPostprocessorInputs::Vector<dim> &input_data,
                                     std::vector<Vector<double>> &computed_quantities) const;
};

template <int dim>
void HfieldPostprocessor<dim>::evaluate_vector_field(const DataPostprocessorInputs::Vector<dim> &input_data,
                                                     std::vector<Vector<double>> &computed_quantities) const {
  // Ensure that there really are as many output slots as there are points at which DataOut provides the curl values:
  AssertDimension(input_data.solution_gradients.size(), computed_quantities.size());
  for (unsigned int p = 0; p < input_data.solution_gradients.size(); ++p) {
    // Ensure that each output slot has exactly 'dim' components (as should be expected, given that we want to create
    // vector-valued outputs), and copy the curl of the solution at the evaluation points into the output slots:
    AssertDimension(computed_quantities[p].size(), (Tensor<1, dim>::n_independent_components));
    for (unsigned int i = 0; i < dim; ++i) {
      double h_field_i_component = 0;
      switch (i) {
      case 0:
        h_field_i_component = input_data.solution_gradients[p][2][1] - input_data.solution_gradients[p][1][2];
        break;
      case 1:
        h_field_i_component = input_data.solution_gradients[p][0][2] - input_data.solution_gradients[p][2][0];
        break;
      case 2:
        h_field_i_component = input_data.solution_gradients[p][1][0] - input_data.solution_gradients[p][0][1];
        break;
      default:
        Assert(false, ExcNotImplemented());
      }
      computed_quantities[p][Tensor<1, dim>::component_to_unrolled_index(TableIndices<1>(i))] = h_field_i_component;
    }
  }
}
```

**Comment by dangars** [comment]
Right now the current function for DataPostprocessor is:
```c++
add_data_vector (const VectorType &data,
const DataPostprocessor< DoFHandlerType::space_dimension > &data_postprocessor)
```

I guess that the solution to this issue would be to add the following function:
```c++
add_data_vector (const VectorType &data,
const DataPostprocessor< DoFHandlerType::space_dimension > &data_postprocessor,
const std::vector< std::string > &names,
const std::vector< DataComponentInterpretation::DataComponentInterpretation > 
&data_component_interpretation)
```

If this is a good solucion, let me know and I can write the PR, otherwise which would be the solution?

**Comment by dangars** [comment]
Another issue is that `DataOut` does not allow to put a complex vector in two separate vectors one for the real part and another one for the imaginary part. For example for the vector `electric_field` I get:
``` c++
"electric_field_x_re", "electric_field_x_im", "electric_field_y_re", 
"electric_field_y_im", "electric_field_z_re", "electric_field_z_im"
```
If I have several eigenmodes that makes a huge list in Paraview. It would be better to have just two vectors per complex vector. One for the real part and another one for the imaginary part. That way I can do vector field representation, etc. Right now I reconstruct the vector using the calculator filter of Paraview.

I don't know how to address this issue. Do you have a hint?

**Comment by dangars** [comment]
Thinking about `DataOut`, I realized that another issue is that the separation is done before `DataPostprocessor`, that means that it is not possible to do complex operations. For example the magnetic field is `h_field = i*curl(E)/(omega*mu)`, although I can always output `curl(E)` with deal.II and then calculate the magnetic field with Paraview.

**Comment by dangars** [comment]
@bangerth You can find the test in #8657, I think that #8648 does not solve the problem.

I managed to calculate `H=i*curl(E)/(omega*mu)` but I had to separate the solution in real and imaginary before DataPostprocessor and DataOut. Then I tell `DataProcessor` if I'm giving him the real or the imaginary part. I can do this because Re(H) depends only on Im(E) and Im(H) depends only on Re(E).

The problem is that `Re(S) = Re((1/2) E^H*)` depends on all the components of Re(E) and all the components of Im(E). It would be possible to use the approach of #8650 if you separate the real and the imaginary parts for the whole calculation (like in step-29), because DataPostprocessor has access to all the components real and imaginary at the same time. That is not the case in the current implementation of std::complex DataPostprocessor, because you have only access to the real or to the imaginary part, but you don't have access to both at the same time. In addition it is not possible to know in DataPostprocessor if you are evaluating the real or the imaginary part.

I took a closer look to the implementation of write_vtu_parallel() and write_hdf5_parallel(). I wonder if it would be better to leave the type conversion (complex->double) as an "implementation detail" and do the type conversion at the very end.

For example, hdf5 is very efficient with the following operation: HDF5 can take a buffer with `std::complex` data and write the real data in one dataset and the imaginary data in another dataset. I would use hpyerslabs for this operation.

In the case of vtu, I suspect that the operation would be less efficient and it will require to re-copy the data in two buffers one for the real data and another one for the imaginary data.

This probably would be mean that `data_sets` from `DataOutFilter` would be `std::vector<std::vector<number>>` where number can be `double` or `std::complex`,

I see that there is already lot of work that you have done to make DataOut compatible with std::complex #1894, #5333, #5335, #5335, #5419. Unfortunately the approach that I propose is orthogonal to those commits.

I would be happy to work on this, as I'm very interested in complex-valued arithmetic with DataPostprocessorVector and I would like to save a complex vector as a a real vector and an imaginary vector.

Let me know if you think if this approach makes sense  :)

**Comment by bangerth** [comment]
@dangars -- should finally be fixed by #9815.
